From b71beb5d9fbd0cc22e696171b23e0e1cb2013add Mon Sep 17 00:00:00 2001
From: Sharaf Zaman <shzam@sdf.org>
Date: Tue, 8 Feb 2022 10:15:13 +0000
Subject: [PATCH] Android: Reset backing store connection with window

When the platform plugin is destroyed. This enables the connection to be
restablished.
---
 .../platforms/android/qandroidplatformbackingstore.h |  2 ++
 .../platforms/android/qandroidplatformwindow.cpp     | 12 ++++++++++++
 .../platforms/android/qandroidplatformwindow.h       |  1 +
 3 files changed, 15 insertions(+)

diff --git a/src/plugins/platforms/android/qandroidplatformbackingstore.h b/src/plugins/platforms/android/qandroidplatformbackingstore.h
index ec1761c789..b0b2c71cf0 100644
--- a/src/plugins/platforms/android/qandroidplatformbackingstore.h
+++ b/src/plugins/platforms/android/qandroidplatformbackingstore.h
@@ -55,6 +55,8 @@ public:
     void resize(const QSize &size, const QRegion &staticContents) override;
     QImage toImage() const override;
     void setBackingStore(QWindow *window);
+    void unsetBackingStoreConnection() { m_backingStoreSet = false; }
+
 protected:
     QImage m_image;
     bool m_backingStoreSet = false;
diff --git a/src/plugins/platforms/android/qandroidplatformwindow.cpp b/src/plugins/platforms/android/qandroidplatformwindow.cpp
index 1574b8b013..be50fa4854 100644
--- a/src/plugins/platforms/android/qandroidplatformwindow.cpp
+++ b/src/plugins/platforms/android/qandroidplatformwindow.cpp
@@ -44,6 +44,7 @@
 
 #include "androidjnimain.h"
 
+#include "qandroidplatformbackingstore.h"
 #include <qguiapplication.h>
 #include <qpa/qwindowsysteminterface.h>
 
@@ -59,6 +60,17 @@ QAndroidPlatformWindow::QAndroidPlatformWindow(QWindow *window)
     setWindowState(window->windowStates());
 }
 
+QAndroidPlatformWindow::~QAndroidPlatformWindow()
+{
+    // Backing store may out live us and keeping the backing store pointed to a
+    // destroyed window causes problems, like window not being rendered at all
+    // (because the backing store wouldn't know to reattach itself to the new
+    // platformWindow).
+    if (m_backingStore) {
+        m_backingStore->unsetBackingStoreConnection();
+    }
+}
+
 void QAndroidPlatformWindow::lower()
 {
     platformScreen()->lower(this);
diff --git a/src/plugins/platforms/android/qandroidplatformwindow.h b/src/plugins/platforms/android/qandroidplatformwindow.h
index aad001c359..0ed0e71890 100644
--- a/src/plugins/platforms/android/qandroidplatformwindow.h
+++ b/src/plugins/platforms/android/qandroidplatformwindow.h
@@ -53,6 +53,7 @@ class QAndroidPlatformWindow: public QPlatformWindow
 {
 public:
     explicit QAndroidPlatformWindow(QWindow *window);
+    ~QAndroidPlatformWindow();
 
     void lower() override;
     void raise() override;
-- 
2.34.1

